### PR TITLE
On recovery: re-check children

### PR DESCRIPTION
**Pull request by Al2Klimov**

### CODE REVIEW

**Comment by Al2Klimov** [comment]
ref/NC/690718

**Review by lippserd — CHANGES_REQUESTED** [review]
Please include a test protocol.

**Comment by Al2Klimov** [comment]
```
➜  icinga2 git:(bugfix/children-recover-too-late) curl -fksSLu root:icinga -H 'Accept: application/json' -X PUT 'https://localhost:5665/v1/objects/hosts/firewall' -d '{ "attrs": { "check_command": "passive", "enable_active_checks": false }, "pretty": true }'; echo
{
    "results": [
        {
            "code": 200.0,
            "status": "Object was created"
        }
    ]
}
➜  icinga2 git:(bugfix/children-recover-too-late) curl -fksSLu root:icinga -H 'Accept: application/json' -X PUT 'https://localhost:5665/v1/objects/hosts/switch' -d '{ "attrs": { "check_command": "disk", "check_interval": 86400 }, "pretty": true }'; echo
{
    "results": [
        {
            "code": 200.0,
            "status": "Object was created"
        }
    ]
}
➜  icinga2 git:(bugfix/children-recover-too-late) curl -fksSLu root:icinga -H 'Accept: application/json' -X PUT 'https://localhost:5665/v1/objects/dependencies/switch!firewall' -d '{ "attrs": { "parent_host_name": "firewall", "disable_checks": true }, "pretty": true }'; echo
{
    "results": [
        {
            "code": 200.0,
            "status": "Object was created"
        }
    ]
}
➜  icinga2 git:(bugfix/children-recover-too-late) for i in {1..3}; do curl -fksSLu root:icinga -H 'Accept: application/json' -X POST 'https://localhost:5665/v1/actions/process-check-result' -d '{ "type": "Host", "filter": "host.name == {{{firewall}}}", "exit_status": 1, "plugin_output": " ", "pretty": true }'; echo; done
{
    "results": [
        {
            "code": 200.0,
            "status": "Successfully processed check result for object 'firewall'."
        }
    ]
}
{
    "results": [
        {
            "code": 200.0,
            "status": "Successfully processed check result for object 'firewall'."
        }
    ]
}
{
    "results": [
        {
            "code": 200.0,
            "status": "Successfully processed check result for object 'firewall'."
        }
    ]
}
➜  icinga2 git:(bugfix/children-recover-too-late) curl -fksSLu root:icinga -H 'Accept: application/json' -X POST 'https://localhost:5665/v1/actions/process-check-result' -d '{ "type": "Host", "filter": "host.name == {{{firewall}}}", "exit_status": 0, "plugin_output": " ", "pretty": true }'; echo
{
    "results": [
        {
            "code": 200.0,
            "status": "Successfully processed check result for object 'firewall'."
        }
    ]
}
➜  icinga2 git:(bugfix/children-recover-too-late)
```

```
[2020-11-10 11:52:45 +0100] information/ApiListener: New client connection from [::1]:60863 (no client certificate)
[2020-11-10 11:52:45 +0100] information/HttpServerConnection: Request: POST /v1/actions/process-check-result (from [::1]:60863), user: root, agent: curl/7.64.1).
[2020-11-10 11:52:45 +0100] information/HttpServerConnection: HTTP client disconnected (from [::1]:60863)
[2020-11-10 11:52:45 +0100] information/ApiListener: New client connection from [::1]:60864 (no client certificate)
[2020-11-10 11:52:45 +0100] information/HttpServerConnection: Request: POST /v1/actions/process-check-result (from [::1]:60864), user: root, agent: curl/7.64.1).
[2020-11-10 11:52:45 +0100] information/HttpServerConnection: HTTP client disconnected (from [::1]:60864)
[2020-11-10 11:52:45 +0100] information/ApiListener: New client connection from [::1]:60865 (no client certificate)
[2020-11-10 11:52:45 +0100] information/HttpServerConnection: Request: POST /v1/actions/process-check-result (from [::1]:60865), user: root, agent: curl/7.64.1).
[2020-11-10 11:52:45 +0100] information/HttpServerConnection: HTTP client disconnected (from [::1]:60865)
[2020-11-10 11:52:48 +0100] information/ApiListener: New client connection from [::1]:60867 (no client certificate)
[2020-11-10 11:52:48 +0100] information/HttpServerConnection: Request: POST /v1/actions/process-check-result (from [::1]:60867), user: root, agent: curl/7.64.1).
[2020-11-10 11:52:48 +0100] information/HttpServerConnection: HTTP client disconnected (from [::1]:60867)
[2020-11-10 11:52:58 +0100] warning/PluginCheckTask: Check command for object 'switch' (PID: 83069, arguments: '/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_disk' '-c' '10%' '-w' '20%' '-X' 'none' '-X' 'tmpfs' '-X' 'sysfs' '-X' 'proc' '-X' 'configfs' '-X' 'devtmpfs' '-X' 'devfs' '-X' 'mtmfs' '-X' 'tracefs' '-X' 'cgroup' '-X' 'fuse.gvfsd-fuse' '-X' 'fuse.gvfs-fuse-daemon' '-X' 'fdescfs' '-X' 'overlay' '-X' 'nsfs' '-X' 'squashfs' '-m') terminated with exit code 128, output: execvpe(/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_disk) failed: No such file or directory
```

File: lib/icinga/checkable-check.cpp [hunk 1]
```diff
@@ -242,6 +242,16 @@ void Checkable::ProcessCheckResult(const CheckResult::Ptr& cr, const MessageOrig
 			OnReachabilityChanged(this, cr, children, origin);
 	}
 
+	if (recovery) {
+		// W/o this children could recover too late, after their problem notifications have been sent.
+		for (auto& child : children) {
+			if (child->GetProblem() && child->GetEnableActiveChecks() && !child->IsLikelyToBeCheckedSoon()) {
+				ObjectLock oLock (child);
+				child->SetNextCheck(now + 10);
```
Comment (lippserd) [inline lib/icinga/checkable-check.cpp:250]: Why `+ 10`?

Comment (Al2Klimov) [reply lib/icinga/checkable-check.cpp:250]: Just to be sure we have enough time to finish processing the current check result.

**Review by lippserd — APPROVED** [review]

**Review by lippserd — APPROVED** [review]

**Review by N-o-X — APPROVED** [review]

**Pull request merged**